src/html-imports.js

```diff
@@ -9,718 +9,780 @@
+          for (let j = 0, ll = s.attributes.length; j < ll; j++) {
+            clone.setAttribute(s.attributes[j].name, s.attributes[j].value);
+          }
+          clone.textContent = s.textContent;
+
+          // Update currentScript and replace original with clone script.
+          currentScript = clone;
+          s.parentNode.replaceChild(clone, s);
+          // Wait for load/error events; after is loaded, reset currentScript.
+          return whenElementLoaded(clone).then(() => currentScript = null);
         });
       }
+      return promise;
     }
-  }
-  function loadedImport(e) {
-    markTargetLoaded(e);
-    newImports.push(this);
-    parsedCount++;
-    checkDone();
-  }
-  function errorLoadingImport(e) {
-    errorImports.push(this);
-    parsedCount++;
-    checkDone();
-  }
-  if (importCount) {
-    for (var i=0, imp; i<importCount && (imp=imports[i]); i++) {
-      if (isImportLoaded(imp)) {
-        newImports.push(this);
-        parsedCount++;
-        checkDone();
-      } else {
-        imp.addEventListener('load', loadedImport);
-        imp.addEventListener('error', errorLoadingImport);
+
+    /**
+     * Waits for all the imported stylesheets/styles to be loaded.
+     * @return {Promise}
+     */
+    _waitForStyles() {
+      // <link rel=stylesheet> should be appended to <head>. Not doing so
+      // in IE/Edge breaks the cascading order
+      // https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/10472273/
+      // If there is one <link rel=stylesheet> imported, we must move all imported
+      // links and styles to <head>.
+      const needsMove = !!document.querySelector(disabledLinkSelector);
+      const s$ = /** @type {!NodeList<!(HTMLLinkElement|HTMLStyleElement)>} */
+        (document.querySelectorAll(pendingStylesSelector));
+      const promises = [];
+      for (let i = 0, l = s$.length, s; i < l && (s = s$[i]); i++) {
+        // Listen for load/error events, remove selector once is done loading.
+        promises.push(whenElementLoaded(s)
+          .then(() => s.removeAttribute(importDependencyAttr)));
+        // Check if was already moved to head, to handle the case where the element
+        // has already been moved but it is still loading.
+        if (needsMove && s.parentNode !== document.head) {
+          let rootImport = importForElement(s);
+          while (rootImport && importForElement(rootImport)) {
+            rootImport = importForElement(rootImport);
+          }
+          // Replace the element we're about to move with a placeholder
+          // containing a reference to the moved element.
+          const placeholder = document.createElement(s.localName);
+          placeholder['__appliedElement'] = s;
+          s.parentNode.replaceChild(placeholder, s);
+          if (rootImport.parentNode === document.head) {
+            document.head.insertBefore(s, rootImport);
+          } else {
+            document.head.appendChild(s);
```

Comment:
Can these not go in `<body>`?

Moving these to head has the potential to break cascade order:
```html
<html>
  <head></head>
  <body>
    <style>/*...*/</style>
    <link rel="import" href="styles-will-be-moved-into-head.html">
  </body>
</html>
```